docs: new feature custom transaction metadata (#1021)
neo4j-contrib/neo4j-spark-connector: docs/modules/ROOT/pages/configuration.adoc

[#options]
= Options and configuration
:description: The chapter describes configuration options of Neo4j Connector for Apache Spark.

Besides using spark options to control how the connector reads and writes to and from Neo4j, there are some extended use cases.

> Here in this page you can read about how to use the spark connector to manage multiple connections and Neo4j driver options and more.

Reviewer: ```suggestion
Spark options are useful not only to control how the connector reads to and writes from a Neo4j database, but also to manage multiple connections and manage Neo4j driver options.
```

== Multiple connections

Neo4j Connector for Apache Spark allows you to use more than one connection in a single Spark Session.

> For example, you can read data from a database and write them in another database in the same session.

Reviewer: ```suggestion
The connector allows you to use more than one connection within a single Spark Session.
For example, you can read data from a database and write it to another database in the same session.
```

  .load()
----

== Neo4j driver options

Under the covers, the Spark connector uses the link:https://neo4j.com/docs/java-manual/current/get-started/#java-driver-get-started-about[official Neo4j Java Driver].
In many situations, you want the control to set driver options to account for your production deployment of Neo4j and how to communicate with it.

When using the connector, any valid Neo4j driver option can be set using the `option` method in

> Spark, like so:

Reviewer: ```suggestion
When using the connector, you can set any valid Neo4j driver option using the `option` method:
```

[source,scala]
----
import org.apache.spark.sql.{SaveMode, SparkSession}

val spark = SparkSession.builder().getOrCreate()

val df = spark.read.format("org.neo4j.spark.DataSource")
        .option("url", "neo4j://localhost:7687")
        .option("authentication.type", "basic")
        .option("authentication.basic.username", "myuser")
        .option("authentication.basic.password", "neo4jpassword")
        .option("labels", "Person")
        .load()
----

Alternatively, you can specify a global configuration in the Spark Session to avoid retyping connection options every time.

> You can set any Neo4j Connector option, just prepend it with `neo4j.`.

Reviewer: ```suggestion
You can set any Neo4j Connector option adding the `neo4j.` prefix.
```

> For example, if you want to set the option `authentication.type` in the session, you have to type `neo4j.authentication.type`.

Reviewer: ```suggestion
For example, if you want to set the `authentication.type` option for the session, you have to add `neo4j.authentication.type`.
```

|No
|===

[role=label--new-6.0]
== Driver transaction metadata

If you for any reason want to label your transactions that are ultimately being sent to Neo4j you can do so via transaction metadata.

> Any transaction metadata that you set will be visible in your Neo4j's query log, making it easy to identify what incoming cypher query originates from what Spark operation.

Reviewer: ```suggestion
You can use transaction metadata to label your transactions in the target Neo4j database query log, to help identifying them as Cypher queries generated by the Spark connector.
```

Metadata is sent as a object/map in the driver, and you populate it using options prefix `db.transaction.metadata`.

> Keys are dot-separated and interpreted right after the prefix, like so

Reviewer: ```suggestion
Keys are dot-separated and interpreted right after the prefix:
```

[.tabbed-example]
====
[.include-with-Scala]
=====
.Example
[source, scala, role=nocollapse]
----
include::example$scala/TransactionMetadata.scala[tags=code-single-metadata]
----
=====

[.include-with-Python]
=====
.Example
[source, python, role=nocollapse]
----
include::example$python/transaction_metadata.py[tags=code-single-metadata]
----
=====
====

> Will send this object in the driver via transaction metadata:

Reviewer: ```suggestion
.Send the object to the driver via transaction metadata.
```

[source, json, role=nocollapse]
include::example$transaction-metadata-single.json[]

> You can created nested structures by simply assigning more options.

Reviewer: ```suggestion
You can create nested structures by assigning more options.
```

Each dot represents one step deeper into the nested map/object.

[.tabbed-example]
====
[.include-with-Scala]
=====
.Example
[source, scala, role=nocollapse]
----
include::example$scala/TransactionMetadata.scala[tags=code-nested-metadata]
----
=====

[.include-with-Python]
=====
.Example
[source, python, role=nocollapse]
----
include::example$python/transaction_metadata.py[tags=code-nested-metadata]
----
=====
====

> Will send this object in the driver via transaction metadata:

Reviewer: ```suggestion
.Send the object to the driver via transaction metadata.
```
